### ISSUE TITLE
[BUG] Segfault in __pyx_convert_string_from_py_std__in_string(object*) when making a struct with a struct member whose members are of type unsigned int, string, string.

### ISSUE THREAD

**Issue by gamedev33** [cython/cython#5447]
### Describe the bug

I used Valgrind to track down this segfault happening practically immediately upon program execution, before a single print can be executed (the first lines are just imports and includes).
The segfault happens in __pyx_convert_string_from_py_std__in_string(_object*).
I tried posting the cpp file generated to pastebin but my browser freezes due to too many lines.

Here's the code block associated with __pyx_convert_string_from_py_std__in_string(object*):
```c++
/* "string.from_py":13
 * 
 * @cname("__pyx_convert_string_from_py_std__in_string")
 * cdef string __pyx_convert_string_from_py_std__in_string(object o) except *:             # <<<<<<<<<<<<<<
 *     cdef Py_ssize_t length = 0
 *     cdef const char* data = __Pyx_PyObject_AsStringAndSize(o, &length)
 */

static std::string __pyx_convert_string_from_py_std__in_string(PyObject *__pyx_v_o) {
  Py_ssize_t __pyx_v_length;
  char const *__pyx_v_data;
  std::string __pyx_r;
  __Pyx_RefNannyDeclarations
  char const *__pyx_t_1;
  std::string __pyx_t_2;
  int __pyx_lineno = 0;
  const char *__pyx_filename = NULL;
  int __pyx_clineno = 0;
  __Pyx_RefNannySetupContext("__pyx_convert_string_from_py_std__in_string", 0);
```

Here's the structs in C++:
```c++
struct CreatureSummary1 {
  unsigned int cid;
  std::string article;
  std::string name;
};

struct Condition {
  long long original;
  long long value;
  CreatureSummary1 actor;
  double end;
```

Without the CreatureSummary1 actor; line/member the program works fine.
This is imported into Cython using cdef extern from "declarations.h": ...

I tried reproducing in just a few lines of code, but I couldn't, so there must be some complications going on.

Standalone.cpp can be downloaded here: https://ufile.io/6p05tdwm
The crash happens at line 13242.

### Code to reproduce the behaviour:

Can't reproduce in a few lines of code.

### OS

Debian 11

### Python version

3.9.2

### Cython version

3.0.0b2

### Additional context

Also happens in Cython version 3.0.0a11.

**Comment by da-woods** [comment]
It'd probably be most use to post a pyx file that can reproduce the issue. Obviously the C++ files are large and unwieldy, but we can generate those ourselves

**Comment by gamedev33** [comment]
> It'd probably be most use to post a pyx file that can reproduce the issue. Obviously the C++ files are large and unwieldy, but we can generate those ourselves

Unfortunately there's like 15'000 lines of code and it's all proprietary code.
I can only post code segments, or I can invite specific members like yourself to the repo with all the code with only the license to read, not write or duplicate code from the repo except very small/modified segments to reproduce the bug isolated from the rest of the code if possible.
But like I said if I comment out the CreatureSummary1 actor member, then there's no segfault.

**Comment by da-woods** [comment]
I don't think this is something I can reasonably debug so I'm going to take a wild guess:

You might have got a global, Cython-defined, c++ class, that causes string conversion to happen in it's constructor.  It's essentially this bug: https://github.com/cython/cython/pull/4228/files#diff-c0209e949522f6c57ec3d884419ba2251e497f6794b705c5384d1a6df9449627 - a lot of Cython code relies on a certain amount of the module to be initialized (e.g. string constants for example), but such a global constructor would end up happening before any of that is initialized.

I'd say ~50% chance that's your issue. But not in a position to find that out definitively.

**Comment by scoder** [comment]
> Here's the code block associated with __pyx_convert_string_from_py_std__in_string(object*):

What you copied is not the complete function but just a boring initial part of it. I doubt that it would be helpful, though, since the function itself is likely ok. The stack trace of the crash would probably be more interesting.

**Comment by gamedev33** [comment]
> I don't think this is something I can reasonably debug so I'm going to take a wild guess:
> 
> You might have got a global, Cython-defined, c++ class, that causes string conversion to happen in it's constructor. It's essentially this bug: https://github.com/cython/cython/pull/4228/files#diff-c0209e949522f6c57ec3d884419ba2251e497f6794b705c5384d1a6df9449627 - a lot of Cython code relies on a certain amount of the module to be initialized (e.g. string constants for example), but such a global constructor would end up happening before any of that is initialized.
> 
> I'd say ~50% chance that's your issue. But not in a position to find that out definitively.

You're right, that's probably the issue. I forgot about this problem.
Thank you. I will figure out a temporary solution, like heap allocation or calling that piece of code after the constructor.

> > Here's the code block associated with __pyx_convert_string_from_py_std__in_string(object*):
> 
> What you copied is not the complete function but just a boring initial part of it. I doubt that it would be helpful, though, since the function itself is likely ok. The stack trace of the crash would probably be more interesting.

Unfortunately that was the entire stack trace. Here's the complete listing in its entirety:
```
Process terminating with default action of signal 11 (SIGSEGV)
==9998==  Access not within mapped region at address 0x8
==9998==    at 0x158C7B: __pyx_convert_string_from_py_std__in_string(_object*) (standalone.cpp:130597)
==9998==  If you believe this happened as a result of a stack
==9998==  overflow in your program's main thread (unlikely but
==9998==  possible), you can try to increase the size of the
==9998==  main thread stack using the --main-stacksize= flag.
==9998==  The main thread stack size used in this run was 8388608.
==9998==
==9998== Events    : Ir
==9998== Collected : 0
==9998==
==9998== I   refs:      0
```

**Comment by gamedev33** [comment]
Update: Heap allocation didn't work. Getting the same segfault at the same time with the same stack trace from valgrind.
But interestingly heap allocation works in a simple test when stack allocation does not (segfault).... hmmm....
